### PR TITLE
Add use statement

### PR DESCRIPTION
**Pull request by cmester0**
Add `use` statements (with resolution information). Imports in coq still need some generalization (currently assume imports are in the `Examples` package), and the updated coq library for hacspec needs to be put somewhere.

### CODE REVIEW

**Comment by cmester0** [comment]
Implements part of the coq backend for #74.

**Comment by W95Psp** [comment]
Thanks for the PR! :)

I'm looking into this, and I'm wondering: do you need `use` items because you actually need to generate Requires in Coq? (i.e. otherwise modules are not in scope)

In this case, `use` items is not what you want I think. I mean, in Rust one might just write fully qualified paths without any use.
I think we should have a phase/util that browses an AST and lists all the items used that live outside of the current crate.

Also, it seems that at the HIR/THIR phase, since name resolution is already there, every use .. as .. items are just gone :(

**Comment by cmester0** [comment]
I think the information is in the resolution objects that are part of `UsePath`s? There at least you can check that it is a `use` statement.

**Comment by W95Psp** [comment]
Well, what I mean is that all user-defined naming is gone at HIR/THIR time.
So, when you do `use foo::bar as hello`, at HIR we actually get `use foo::bar`. 
This is a bit of an issue, because we usually do renaming when there are name clashes or to make things clearer.

But anyway, these kinds of issues are not very important, it's only about making paths nicer.

The important thing is: do you need the list of external crate used in an AST so that you can generate `Require`s in Coq? (because in this case, the current phases / utils / AST doesn't give enough information indeed!)

**Comment by cmester0** [comment]
I found a way to get the `rename` information back, it is stored as the ident on item objects representing `use` statements (https://github.com/rust-lang/rust/blob/8b8110e1469d459a196f6feb60d82dec48c3cfc2/compiler/rustc_ast_lowering/src/item.rs#L563-L569), will try and add it to the path objects.

File: engine/lib/ast.ml [hunk 1]
```diff
@@ -553,6 +553,7 @@ functor
           of_trait : (global_ident * generic_value list) option;
           items : impl_item list;
         }
+      | Use of string list * string * bool * string option
```
Comment (W95Psp) [inline engine/lib/ast.ml:556]: For uniformity, I'd prefer a record payload here, i.e.
```suggestion
      | Use of { path: string list; external: bool; rename: string }
```

About the `string` field for Single, Glob, etc.: I don't really see how we could use this information. Do you have ideas in mind?

**Review by W95Psp — COMMENTED** [review]
just leaving some quick comment (even if it's still draft)

File: engine/lib/ast.ml (hunk 1 above)
Comment (cmester0) [reply engine/lib/ast.ml:556]: I agree, I just did not want to throw away potential useful information. However this can be re-added if it is ever needed.

**Review by W95Psp — APPROVED** [review]
LGTM, thanks! 😃

**Pull request merged**